### ISSUE TITLE
Using C++ Interface on threads causes collector to crash

### ISSUE THREAD

**Issue by comeradealexi** [ivmai/bdwgc#162]
Running on Windows x64.

I've modified the test function in thread_leak_test.c to run the code below. So 5 threads are running the code below simultaneously. 

```
#include <gc_cpp.h>
class A : public gc_cleanup
{

	int iVal;
};

#ifdef GC_PTHREADS
  void * test(void * arg)
#else
  DWORD WINAPI test(LPVOID arg)
#endif
{
	  new A();

#   ifdef GC_PTHREADS
      return arg;
#   else
      return (DWORD)(GC_word)arg;
#   endif
```

The rest of the code in the sample is the same.

I've got the following defined: GC_WIN32_THREADS GC_DEBUG GC_ASSERTIONS.

The crash is isn't always the same but it's often in mark.c on the following line inside GC_is_marked:     `return (int)mark_bit_from_hdr(hhdr, bit_no); /* 0 or 1 */` with an access violation.

Sometimes it will actually work, othertimes it will crash somewhere entirely different. Changing the number of new'd A's in the test function alters the outcome and the type of crash.

Is the gc_cleanup/C++ interface not suitable to be used on threads?

Alex

**Comment by ivmai** [comment]
I have reproduce it (SIGSEGV in GC_is_marked):
gcc -I include -DGC_THREADS -O0 -g -c [-DGC_ASSERTIONS] extra/gc.c
g++ -I include -DGC_THREADS -O0 -g test.c gc.o

test.c:
#undef GC_NO_THREAD_REDIRECTS

#ifdef GC_PTHREADS
# include <pthread.h>
#else
# include <windows.h>
#endif

#include <stdio.h>

#include <gc_cpp.h>
class A : public gc_cleanup
{

	int iVal;
};

#ifdef GC_PTHREADS
  void * test(void * arg)
#else
  DWORD WINAPI test(LPVOID arg)
#endif
{
	  new A();

#   ifdef GC_PTHREADS
      return arg;
#   else
      return (DWORD)(GC_word)arg;
#   endif
}

#define NTHREADS 5

int main(void) {
    int i;
#   ifdef GC_PTHREADS
      pthread_t t[NTHREADS];
#   else
      HANDLE t[NTHREADS];
      DWORD thread_id;
#   endif
    int code;

    GC_set_find_leak(1); // it does not crash in non-leak mode
    GC_INIT();

    for (i = 0; i < NTHREADS; ++i) {
#       ifdef GC_PTHREADS
          code = pthread_create(t + i, 0, test, 0);
#       else
          t[i] = CreateThread(NULL, 0, test, 0, 0, &thread_id);
          code = t[i] != NULL ? 0 : (int)GetLastError();
#       endif
        if (code != 0) {
            fprintf(stderr, "Thread creation failed %d\n", code);
            exit(2);
        }
    }

    for (i = 0; i < NTHREADS; ++i) {
#       ifdef GC_PTHREADS
          code = pthread_join(t[i], 0);
#       else
          code = WaitForSingleObject(t[i], INFINITE) == WAIT_OBJECT_0 ? 0 :
                                                        (int)GetLastError();
#       endif
        if (code != 0) {
            fprintf(stderr, "Thread join failed %d\n", code);
            exit(2);
        }
    }

    GC_gcollect();
    GC_gcollect();
    GC_gcollect();
    return 0;
}

**Comment by ivmai** [comment]
Fixed by commit bd4b251
